deploy-agent/deployd/common/stats.py

```diff
@@ -28,69 +49,287 @@ def __exit__(self, stats, sample_rate=1.0, tags=None):
     if IS_PINTEREST:
-        try:
-            from pinstatsd.statsd import sc
-            tags = _add_default_tags(tags)
-            sc.gauge(stat, value, sample_rate=sample_rate, tags=tags)
-        except Exception as e:
-            error_msg = str(e)
-            log.error('unable to send metric: {}'.format(error_msg))
+        mtype = 'gauge'
+        client = MetricClient()
+        client.send(mtype=mtype,
+                    name=name,
+                    value=value,
+                    sample_rate=sample_rate,
+                    tags=tags)
     else:
         return
 
 
+class MetricCache:
+    """ local cache for metrics
+        creates empty cache file
+    """
+    def __init__(self, path=METRIC_CACHE_PATH):
+        self.path = path
+        # maximum cache size in bytes
+        self.max_size = (10 * 1024 * 1024)
+        # init cache
+        if not self.exists():
+            self.truncate()
+
+    def limit(self):
+        """ check to see if cache file has exceeded maximum size
+            return: bool
+        """
+        return os.path.getsize(self.path) > self.max_size
+
+    def exists(self):
+        """ cache file exists and is read/write
+            return: bool
+        """
+        if os.access(self.path, os.F_OK) and \
+           os.access(self.path, os.R_OK) and \
+           os.access(self.path, os.W_OK):
+            return True
+        return False
+
+    def is_empty(self):
+        """ check if the cache not empty
+            return: bool
+        """
+        return not os.stat(self.path).st_size > 0
+
+    def read(self):
+        """ read metrics from cache, then delete
+            return: generator
+        """
+        with open(self.path, 'r') as fh:
+            for line in fh:
+                yield Stat(ins=line)
+
+    def write(self, output):
+        """ write metrics to cache file respecting max cache size
+            appends newline to metric
+        """
+        if self.limit():
+            msg = 'cache file {} has exceeded maximum file size of {}'
+            log.error(msg.format(self.path, self.max_size))
+            return
+        with open(self.path, 'a') as fh:
+            fh.write('{}\n'.format(output))
+
+    def truncate(self):
+        """ purge cache file """
+        with open(self.path, 'w') as fh:
+            fh.truncate()
+
+
+class Stat:
+    """ stat class for simple data management, dataclasses are py3.7+
+        supports all methods for stats
+    """
+
+    def __init__(self, mtype=None, name=None, value=None, sample_rate=None, tags=None, ins=None):
+        self.mtype = mtype
+        self.name = name
+        self.value = value
+        self.sample_rate = sample_rate
+        self.tags = tags
+        self.ins = ins
+        try:
+            self.JSONDecodeError = json.decoder.JSONDecodeError
+        except AttributeError:
+            # python2 support
+            self.JSONDecodeError = ValueError
+
+    def __repr__(self):
```

Comment:
repr should return a string https://docs.python.org/3/reference/datamodel.html#object.__repr__

Reply:
Huge oversight.  Thanks.  I refactored the code to not even use __repr__ which can be problematic for this use case to begin with.